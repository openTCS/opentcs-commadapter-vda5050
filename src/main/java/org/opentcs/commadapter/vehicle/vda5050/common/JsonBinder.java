/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.common;

import static java.util.Objects.requireNonNull;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.SerializationFeature;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;
import java.io.IOException;
import java.util.function.Function;
import javax.annotation.Nonnull;

/**
 * Binds JSON strings to objects and vice versa.
 * Optionally, a filter can be applied before serializing to JSON.
 */
public class JsonBinder {

  /**
   * Maps between objects and their JSON representations.
   */
  private final ObjectMapper objectMapper
      = new ObjectMapper()
          .registerModule(new JavaTimeModule())
          .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS);

  private Function<JsonNode, JsonNode> filter = Function.identity();

  /**
   * Creates a new instance.
   */
  public JsonBinder() {
  }

  /**
   * Sets the filter that will be applied before serializing to JSON.
   *
   * @param filter The filter to be applied.
   */
  public void setFilter(
      @Nonnull
      Function<JsonNode, JsonNode> filter
  ) {
    this.filter = requireNonNull(filter, "filter");
  }

  /**
   * Maps the given JSON string to an object.
   *
   * @param <T> The type of object to map to.
   * @param jsonString The JSON string.
   * @param clazz The type of object to map to.
   * @return The object created from the JSON string.
   * @throws IllegalArgumentException In case there was a problem mapping the given object from
   * JSON.
   */
  public <T> T fromJson(String jsonString, Class<T> clazz)
      throws IllegalArgumentException {
    try {
      return objectMapper.readValue(jsonString, clazz);
    }
    catch (IOException exc) {
      throw new IllegalArgumentException("Could not parse JSON input", exc);
    }
  }

  /**
   * Applies the filter set via {@link #setFilter(Function)} and maps the given
   * object to a JSON string.
   *
   * @param object The object to be mapped.
   * @return The JSON string representation of the object.
   * @throws IllegalArgumentException In case there was a problem mapping the given object to JSON.
   */
  public String toJson(Object object)
      throws IllegalArgumentException {
    try {
      return objectMapper
          .writerWithDefaultPrettyPrinter()
          .writeValueAsString(filter.apply(objectMapper.valueToTree(object)));
    }
    catch (JsonProcessingException exc) {
      throw new IllegalArgumentException("Could not produce JSON output", exc);
    }
  }
}
