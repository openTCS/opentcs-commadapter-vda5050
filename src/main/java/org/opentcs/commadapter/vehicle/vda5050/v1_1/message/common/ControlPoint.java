// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Defines the control points of NURBS.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ControlPoint
    implements
      Serializable {

  /**
   * X coordinate described in the world coordinate system (in m).
   */
  private Double x;
  /**
   * Y coordinate described in the world coordinate system (in m).
   */
  private Double y;
  /**
   * The weight with which this control point pulls on the curve. Range: [0 ... infinity]
   * <p>
   * When not defined, the default will be 1.0.
   */
  private Double weight;
  /**
   * [Optional] Orientation of the AGV on this position of the curve. Range: [-PI ... PI]
   * <p>
   * The orientation is in world coordinates. When not defined the orientation of the AGV will be
   * tangential to the curve.
   */
  private Double orientation;

  @JsonCreator
  public ControlPoint(
      @Nonnull
      @JsonProperty(required = true, value = "x")
      Double x,
      @Nonnull
      @JsonProperty(required = true, value = "y")
      Double y,
      @Nonnull
      @JsonProperty(required = true, value = "weight")
      Double weight
  ) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
    this.weight
        = checkInRange(requireNonNull(weight, "weight"), 0.0, Double.MAX_VALUE, "weight");
  }

  public Double getWeight() {
    return weight;
  }

  public ControlPoint setWeight(
      @Nonnull
      Double weight
  ) {
    this.weight
        = checkInRange(requireNonNull(weight, "weight"), 0.0, Double.MAX_VALUE, "weight");
    return this;
  }

  public Double getX() {
    return x;
  }

  public ControlPoint setX(
      @Nonnull
      Double x
  ) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public ControlPoint setY(
      @Nonnull
      Double y
  ) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  public Double getOrientation() {
    return orientation;
  }

  public ControlPoint setOrientation(Double orientation) {
    if (orientation != null) {
      checkInRange(orientation, -Math.PI, Math.PI, "orientation");
    }
    this.orientation = orientation;
    return this;
  }

  @Override
  public String toString() {
    return "ControlPoint{" + "x=" + x
        + ", y=" + y
        + ", weight=" + weight
        + ", orientation=" + orientation
        + '}';
  }

}
