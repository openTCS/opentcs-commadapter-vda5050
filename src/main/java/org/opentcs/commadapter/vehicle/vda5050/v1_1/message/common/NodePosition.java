/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Defines the position on a map in world coordinates.
 * <p>
 * Each floor has its own map. All maps must use the same project specific global origin.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class NodePosition
    implements
      Serializable {

  /**
   * X-position on the map in reference to the map coordinate system (in m).
   * Precision is up to the specific implementation.
   */
  private Double x;
  /**
   * Y-position on the map in reference to the map coordinate system (in m).
   * Precision is up to the specific implementation.
   */
  private Double y;
  /**
   * [Optional] Orientation of the AGV on the node (in rad). Range: [-PI ... PI]
   * <p>
   * Optional for AGVs that plan their path by themselves.
   * <p>
   * If defined, the AGV has to assume the theta angle on this node. If previous edge disallows
   * rotation, the AGV is to rotate on the node. If following edge has a differing orientation
   * defined but disallows rotation, the AGV is to rotate on the node to the edges desired rotation
   * before entering the edge.
   */
  private Double theta;
  /**
   * [Optional] Indicates how exact an AGV has to drive over a node in order for it to count as
   * traversed.
   * <p>
   * Interpretation of values:
   * <ul>
   * <li>If equals 0: No deviation is allowed. (No deviation means within the normal tolerance of
   * the AGV manufacturer.)</li>
   * <li>If greater than 0: Allowed deviation radius (in m). If the AGV passes a node within the
   * deviation radius, the node is considered to have been traversed.</li>
   * </ul>
   */
  private Double allowedDeviationXY;
  /**
   * [Optional] Indicates how big the deviation of theta angle can be. Range: [0 ... PI]
   * <p>
   * The lowest acceptable angle is {@code <theta> - <allowedDevaitionTheta>} and the highest
   * acceptable angle is {@code <theta> + <allowedDeviationTheta>}.
   */
  private Double allowedDeviationTheta;
  /**
   * Unique identification of the map in which the position is referenced.
   * <p>
   * Each map has the same origin of coordinates. When an AGV uses an elevator, e. g. leading from a
   * departure floor to a target floor, it will disappear off the map of the departure floor and
   * spawn in the related lift node on the map of the target floor.
   */
  private String mapId;
  /**
   * [Optional] Additional information on the map.
   */
  private String mapDescription;

  @JsonCreator
  public NodePosition(
      @Nonnull
      @JsonProperty(required = true, value = "x")
      Double x,
      @Nonnull
      @JsonProperty(required = true, value = "y")
      Double y,
      @Nonnull
      @JsonProperty(required = true, value = "mapId")
      String mapId
  ) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
    this.mapId = requireNonNull(mapId, "mapId");
  }

  public Double getX() {
    return x;
  }

  public NodePosition setX(
      @Nonnull
      Double x
  ) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public NodePosition setY(
      @Nonnull
      Double y
  ) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  public Double getTheta() {
    return theta;
  }

  public NodePosition setTheta(Double theta) {
    if (theta != null) {
      checkInRange(theta, -Math.PI, Math.PI, "theta");
    }
    this.theta = theta;
    return this;
  }

  public Double getAllowedDeviationXY() {
    return allowedDeviationXY;
  }

  public NodePosition setAllowedDeviationXY(Double allowedDeviationXY) {
    this.allowedDeviationXY = allowedDeviationXY;
    return this;
  }

  public Double getAllowedDeviationTheta() {
    return allowedDeviationTheta;
  }

  public NodePosition setAllowedDeviationTheta(Double allowedDeviationTheta) {
    if (allowedDeviationTheta != null) {
      checkInRange(allowedDeviationTheta, 0, Math.PI, "allowedDeviationTheta");
    }
    this.allowedDeviationTheta = allowedDeviationTheta;
    return this;
  }

  public String getMapId() {
    return mapId;
  }

  public NodePosition setMapId(
      @Nonnull
      String mapId
  ) {
    this.mapId = requireNonNull(mapId, "mapId");
    return this;
  }

  public String getMapDescription() {
    return mapDescription;
  }

  public NodePosition setMapDescription(String mapDescription) {
    this.mapDescription = mapDescription;
    return this;
  }

  @Override
  public String toString() {
    return "NodePosition{" + "x=" + x
        + ", y=" + y
        + ", theta=" + theta
        + ", allowedDeviationXY=" + allowedDeviationXY
        + ", allowedDeviationTheta=" + allowedDeviationTheta
        + ", mapId=" + mapId
        + ", mapDescription=" + mapDescription
        + '}';
  }

}
