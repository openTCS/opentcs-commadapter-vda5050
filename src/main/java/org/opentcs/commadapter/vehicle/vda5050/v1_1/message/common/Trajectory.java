/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import java.util.List;
import javax.annotation.Nonnull;

/**
 * The trajectory of an AGV described as NURBS.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class Trajectory
    implements
      Serializable {

  /**
   * Defines the number of control points that influence any given point on the curve.
   * Range: [1 ... infinity]
   * <p>
   * Increasing the degree increases continuity. If not defined, the default value is 1.
   */
  private Double degree;
  /**
   * Sequence of parameter values that determine where and how the control points affect the NURBS
   * curve. Range: [0.0 ... 1.0]
   * <p>
   * Has size of {@code <number of control points> + <degree> + 1}.
   */
  private List<Double> knotVector;
  /**
   * List of {@link ControlPoint} objects defining the control points of the NURBS, which include
   * the start and end point.
   */
  private List<ControlPoint> controlPoints;

  // CHECKSTYLE:OFF (Long lines because some of these parameter declarations are very long.)
  @JsonCreator
  public Trajectory(
      @Nonnull
      @JsonProperty(required = true, value = "degree")
      Double degree,
      @Nonnull
      @JsonProperty(required = true, value = "knotVector")
      List<Double> knotVector,
      @Nonnull
      @JsonProperty(required = true, value = "controlPoints")
      List<ControlPoint> controlPoints
  ) {
    this.degree
        = checkInRange(requireNonNull(degree, "degree"), 1.0, Double.MAX_VALUE, "degree");
    requireNonNull(knotVector, "knotVector");
    knotVector.forEach(value -> {
      requireNonNull(value, "knotVector value");
      checkInRange(value, 0.0, 1.0, "knotVector value");
    });
    this.knotVector = knotVector;
    this.controlPoints = requireNonNull(controlPoints, "controlPoints");
  }
  // CHECKSTYLE:ON

  public Double getDegree() {
    return degree;
  }

  public Trajectory setDegree(
      @Nonnull
      Double degree
  ) {
    this.degree
        = checkInRange(requireNonNull(degree, "degree"), 1.0, Double.MAX_VALUE, "degree");
    return this;
  }

  public List<Double> getKnotVector() {
    return knotVector;
  }

  public Trajectory setKnotVector(
      @Nonnull
      List<Double> knotVector
  ) {
    requireNonNull(knotVector, "knotVector");
    knotVector.forEach(value -> {
      requireNonNull(value, "knotVector value");
      checkInRange(value, 0.0, 1.0, "knotVector value");
    });
    this.knotVector = knotVector;
    return this;
  }

  public List<ControlPoint> getControlPoints() {
    return controlPoints;
  }

  public Trajectory setControlPoints(
      @Nonnull
      List<ControlPoint> controlPoints
  ) {
    this.controlPoints = requireNonNull(controlPoints, "controlPoints");
    return this;
  }

  @Override
  public String toString() {
    return "Trajectory{" + "degree=" + degree
        + ", knotVector=" + knotVector
        + ", controlPoints=" + controlPoints
        + '}';
  }

}
