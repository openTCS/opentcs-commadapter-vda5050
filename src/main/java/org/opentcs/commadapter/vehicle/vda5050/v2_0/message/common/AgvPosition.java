/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common;

import static java.util.Objects.requireNonNull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import javax.annotation.Nonnull;

/**
 * Defines the position on a map in world coordinates. Each floor has its own map.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class AgvPosition
    implements
      Serializable {

  /**
   * X-position on the map in reference to the map coordinate system.
   * <p>
   * Precision is up to the specific implementation.
   */
  private Double x;
  /**
   * Y-position on the map in reference to the map coordinate system.
   * <p>
   * Precision is up to the specific implementation.
   */
  private Double y;
  /**
   * Orientation of the AGV (in rad). Range: [-PI ... PI]
   */
  private Double theta;
  /**
   * Unique identification of the map in which the position is referenced.
   * <p>
   * Each map has the same origin of coordinates. When an AGV uses an elevator, e. g. leading from a
   * departure floor to a target floor, it will disappear off the map of the departure floor and
   * spawn in the related lift node on the map of the target floor.
   */
  private String mapId;
  /**
   * [Optional] Additional information on the map.
   */
  private String mapDescription;
  /**
   * Whether the AGV's position is initialized or not.
   */
  private Boolean positionInitialized;
  /**
   * [Optional] Describes the quality of the localization. Range: [0.0 ... 1.0]
   * <p>
   * Can be used e. g. by SLAM-AGVs to describe how accurate the current position information is.
   * <p>
   * Interpretation of values:
   * <ul>
   * <li>0.0: Position unknown</li>
   * <li>1.0: Position known</li>
   * </ul>
   * <p>
   * Optional for vehicles that cannot estimate their localization score. Only for logging and
   * visualization purposes.
   */
  private Double localizationScore;
  /**
   * [Optional] Value for position deviation range (in m).
   * <p>
   * Optional for vehicles that cannot estimate their deviation (e.g. with grid-based localization).
   * Only for logging and visualization purposes.
   */
  private Double deviationRange;

  // CHECKSTYLE:OFF (Long lines because some of these parameter declarations are very long.)
  @JsonCreator
  public AgvPosition(
      @Nonnull
      @JsonProperty(required = true, value = "x")
      Double x,
      @Nonnull
      @JsonProperty(required = true, value = "y")
      Double y,
      @Nonnull
      @JsonProperty(required = true, value = "theta")
      Double theta,
      @Nonnull
      @JsonProperty(required = true, value = "mapId")
      String mapId,
      @Nonnull
      @JsonProperty(required = true, value = "positionInitialized")
      Boolean positionInitialized
  ) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
    this.theta = checkInRange(requireNonNull(theta, "theta"), -Math.PI, Math.PI, "theta");
    this.mapId = requireNonNull(mapId, "mapId");
    this.positionInitialized = requireNonNull(positionInitialized, "positionInitialized");
  }
  // CHECKSTYLE:ON

  public Double getX() {
    return x;
  }

  public AgvPosition setX(
      @Nonnull
      Double x
  ) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public AgvPosition setY(
      @Nonnull
      Double y
  ) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  public Double getTheta() {
    return theta;
  }

  public AgvPosition setTheta(
      @Nonnull
      Double theta
  ) {
    this.theta = checkInRange(requireNonNull(theta, "theta"), -Math.PI, Math.PI, "theta");
    return this;
  }

  public String getMapId() {
    return mapId;
  }

  public AgvPosition setMapId(
      @Nonnull
      String mapId
  ) {
    this.mapId = requireNonNull(mapId, "mapId");
    return this;
  }

  public String getMapDescription() {
    return mapDescription;
  }

  public AgvPosition setMapDescription(String mapDescription) {
    this.mapDescription = mapDescription;
    return this;
  }

  public Boolean isPositionInitialized() {
    return positionInitialized;
  }

  public AgvPosition setPositionInitialized(
      @Nonnull
      Boolean positionInitialized
  ) {
    this.positionInitialized = requireNonNull(positionInitialized, "positionInitialized");
    return this;
  }

  public Double getLocalizationScore() {
    return localizationScore;
  }

  public AgvPosition setLocalizationScore(Double localizationScore) {
    if (localizationScore != null) {
      checkInRange(localizationScore, 0.0, 1.0, "localizationScore");
    }
    this.localizationScore = localizationScore;
    return this;
  }

  public Double getDeviationRange() {
    return deviationRange;
  }

  public AgvPosition setDeviationRange(Double deviationRange) {
    this.deviationRange = deviationRange;
    return this;
  }

  @Override
  public String toString() {
    return "AgvPosition{" + "x=" + x
        + ", y=" + y
        + ", theta=" + theta
        + ", mapId=" + mapId
        + ", mapDescription=" + mapDescription
        + ", positionInitialized=" + positionInitialized
        + ", localizationScore=" + localizationScore
        + ", deviationRange=" + deviationRange
        + '}';
  }

}
