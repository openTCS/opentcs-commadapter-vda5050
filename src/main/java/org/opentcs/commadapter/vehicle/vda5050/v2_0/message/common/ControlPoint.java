/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonInclude;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.io.Serializable;
import static java.util.Objects.requireNonNull;
import javax.annotation.Nonnull;
import static org.opentcs.commadapter.vehicle.vda5050.common.Assertions.checkInRange;

/**
 * Defines the control points of NURBS.
 */
@JsonInclude(JsonInclude.Include.NON_NULL)
public class ControlPoint
    implements Serializable {

  /**
   * X coordinate described in the world coordinate system (in m).
   */
  private Double x;
  /**
   * Y coordinate described in the world coordinate system (in m).
   */
  private Double y;
  /**
   * [Optional] The weight with which this control point pulls on the curve. Range: [0 ... infinity]
   * <p>
   * When not defined, the default will be 1.0.
   */
  private Double weight;

  @JsonCreator
  public ControlPoint(
      @Nonnull @JsonProperty(required = true, value = "x") Double x,
      @Nonnull @JsonProperty(required = true, value = "y") Double y) {
    this.x = requireNonNull(x, "x");
    this.y = requireNonNull(y, "y");
  }

  public Double getWeight() {
    return weight;
  }

  public ControlPoint setWeight(Double weight) {
    this.weight = weight == null ? null : checkInRange(weight, 0.0, Double.MAX_VALUE, "weight");
    return this;
  }

  public Double getX() {
    return x;
  }

  public ControlPoint setX(@Nonnull Double x) {
    this.x = requireNonNull(x, "x");
    return this;
  }

  public Double getY() {
    return y;
  }

  public ControlPoint setY(@Nonnull Double y) {
    this.y = requireNonNull(y, "y");
    return this;
  }

  @Override
  public String toString() {
    return "ControlPoint{" + "x=" + x
        + ", y=" + y
        + ", weight=" + weight
        + '}';
  }

}
