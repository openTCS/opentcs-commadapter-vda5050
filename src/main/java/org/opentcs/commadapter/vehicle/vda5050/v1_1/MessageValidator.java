/**
 * Copyright (c) The openTCS Authors.
 *
 * This program is free software and subject to the MIT license. (For details,
 * see the licensing information (LICENSE.txt) you should have received with
 * this copy of the software.)
 */
package org.opentcs.commadapter.vehicle.vda5050.v1_1;

import java.io.InputStreamReader;
import java.util.Map;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonValidator;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.connection.Connection;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.instantactions.InstantActions;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.order.Order;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.state.State;
import org.opentcs.commadapter.vehicle.vda5050.v1_1.message.visualization.Visualization;

/**
 * A {@link JsonValidator} for VDA5050 1.1 messages.
 */
public class MessageValidator
    extends
      JsonValidator {

  /**
   * Creates a new instance.
   *
   * @throws IllegalArgumentException If there was a problem reading the VDA5050 schemas.
   */
  public MessageValidator()
      throws IllegalArgumentException {
    super(
        Map.of(
            Connection.class,
            new InputStreamReader(
                Connection.class.getResourceAsStream(Connection.JSON_SCHEMA_PATH)
            ),
            InstantActions.class,
            new InputStreamReader(
                InstantActions.class.getResourceAsStream(InstantActions.JSON_SCHEMA_PATH)
            ),
            Order.class,
            new InputStreamReader(
                Order.class.getResourceAsStream(Order.JSON_SCHEMA_PATH)
            ),
            State.class,
            new InputStreamReader(
                State.class.getResourceAsStream(State.JSON_SCHEMA_PATH)
            ),
            Visualization.class,
            new InputStreamReader(
                Visualization.class.getResourceAsStream(Visualization.JSON_SCHEMA_PATH)
            )
        )
    );
  }

}
