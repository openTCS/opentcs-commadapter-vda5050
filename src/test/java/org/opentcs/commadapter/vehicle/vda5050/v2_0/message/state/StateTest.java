// SPDX-FileCopyrightText: The openTCS Authors
// SPDX-License-Identifier: MIT
package org.opentcs.commadapter.vehicle.vda5050.v2_0.message.state;

import java.time.Instant;
import java.util.List;
import org.approvaltests.Approvals;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.opentcs.commadapter.vehicle.vda5050.ResourceLoader;
import org.opentcs.commadapter.vehicle.vda5050.common.JsonBinder;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.MessageValidator;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.AgvPosition;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.ControlPoint;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.NodePosition;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.Trajectory;
import org.opentcs.commadapter.vehicle.vda5050.v2_0.message.common.Velocity;

/**
 * Unit tests for {@link State}.
 */
public class StateTest {

  private static final String RESOURCE_DIR
      = "/org/opentcs/commadapter/vehicle/vda5050/v2_0/message/state/";

  private MessageValidator messageValidator;
  private JsonBinder jsonBinder;

  @BeforeEach
  public void setUp() {
    messageValidator = new MessageValidator();
    jsonBinder = new JsonBinder();
  }

  @Test
  public void validateAgainstJsonSchemaMinimal() {
    messageValidator.validate(
        jsonBinder.toJson(createStateMinimal()),
        State.class
    );
  }

  @Test
  public void validateAgainstJsonSchemaFull() {
    messageValidator.validate(
        jsonBinder.toJson(createStateFull()),
        State.class
    );
  }

  @Test
  public void validateJsonWithNullForOptionalFieldsAgainstSchema() {
    messageValidator.validate(
        ResourceLoader.load(RESOURCE_DIR + "stateMessageWithNullForOptionalFields.json"),
        State.class
    );
  }

  @Test
  public void deserializeJsonWithNullForOptionalFields() {
    Approvals.verify(
        jsonBinder.toJson(
            jsonBinder.fromJson(
                ResourceLoader.load(RESOURCE_DIR + "stateMessageWithNullForOptionalFields.json"),
                State.class
            )
        )
    );
  }

  @Test
  public void validateJsonWithNullForOptionalRootFieldsAgainstSchema() {
    messageValidator.validate(
        ResourceLoader.load(RESOURCE_DIR + "stateMessageWithNullForOptionalRootFields.json"),
        State.class
    );
  }

  @Test
  public void deserializeJsonWithNullForOptionalRootFields() {
    Approvals.verify(
        jsonBinder.toJson(
            jsonBinder.fromJson(
                ResourceLoader.load(
                    RESOURCE_DIR + "stateMessageWithNullForOptionalRootFields.json"
                ),
                State.class
            )
        )
    );
  }

  @Test
  public void jsonSampleMinimal() {
    Approvals.verify(jsonBinder.toJson(createStateMinimal()));
  }

  @Test
  public void jsonSampleFull() {
    Approvals.verify(jsonBinder.toJson(createStateFull()));
  }

  private static State createStateMinimal() {
    return new State(
        0L,
        Instant.EPOCH,
        "version",
        "manufacturer",
        "serial-number",
        "some-order",
        0L,
        "some-node",
        0L,
        List.of(),
        List.of(),
        true,
        List.of(),
        new BatteryState(50.0, false),
        OperatingMode.SEMIAUTOMATIC,
        List.of(),
        new SafetyState(EStop.AUTOACK, false)
    );
  }

  private static State createStateFull() {
    return new State(
        0L,
        Instant.EPOCH,
        "version",
        "manufacturer",
        "serial-number",
        "some-order",
        0L,
        "some-node",
        0L,
        List.of(
            new NodeState("some-node", 0L, true)
                .setNodeDescription("some-node-description")
                .setReleased(true)
                .setNodePosition(
                    new NodePosition(1.2, 3.4, "some-map-id")
                        .setTheta(2.7)
                        .setAllowedDeviationXY(0.2)
                        .setAllowedDeviationTheta(0.5)
                        .setMapDescription("some-map-description")
                )
        ),
        List.of(
            new EdgeState("some-edge", 0L, true)
                .setEdgeDescription("some-description")
                .setTrajectory(
                    new Trajectory(
                        1.0,
                        List.of(0.0),
                        List.of(
                            new ControlPoint(9.8, 7.6)
                                .setWeight(2.0)
                        )
                    )
                )
        ),
        true,
        List.of(
            new ActionState("some-action-id", ActionStatus.FAILED)
                .setActionType("some-type")
                .setActionDescription("some-action-description")
                .setResultDescription("some-result-description")
        ),
        new BatteryState(50.0, false)
            .setBatteryVoltage(12.0)
            .setBatteryHealth(70L)
            .setReach(70L),
        OperatingMode.SEMIAUTOMATIC,
        List.of(
            new ErrorEntry(
                "some-error",
                ErrorLevel.FATAL
            )
                .setErrorDescription("some-error-description")
                .setErrorReferences(
                    List.of(
                        new ErrorReference("some-ref-key", "some-ref-value")
                    )
                )
        ),
        new SafetyState(EStop.AUTOACK, false)
    )
        .setZoneSetId("some-zone-set")
        .setAgvPosition(
            new AgvPosition(2.1, 5.4, 0.3, "some-map-id", true)
                .setDeviationRange(12.0)
                .setLocalizationScore(0.5)
                .setMapDescription("some-map-description")
        )
        .setVelocity(
            new Velocity()
                .setVx(0.1)
                .setVy(0.3)
                .setOmega(0.0)
        )
        .setLoads(
            List.of(
                new Load()
                    .setLoadId("some-load")
                    .setLoadType("some-load-type")
                    .setLoadPosition("some-load-position")
                    .setBoundingBoxReference(
                        new BoundingBoxReference(0.0, 0.0, 0.0)
                            .setTheta(0.0)
                    )
                    .setLoadDimensions(
                        new LoadDimensions(0.0, 0.0)
                            .setHeight(1.4)
                    )
                    .setWeight(0L)
            )
        )
        .setInformation(
            List.of(
                new InfoEntry("some-info", InfoLevel.DEBUG)
                    .setInfoDescription("some-info-description")
                    .setInfoReferences(
                        List.of(
                            new InfoReference("some-ref-key", "some-ref-value")
                        )
                    )
            )
        )
        .setPaused(Boolean.FALSE)
        .setNewBaseRequest(Boolean.FALSE)
        .setDistanceSinceLastNode(12.0);
  }
}
