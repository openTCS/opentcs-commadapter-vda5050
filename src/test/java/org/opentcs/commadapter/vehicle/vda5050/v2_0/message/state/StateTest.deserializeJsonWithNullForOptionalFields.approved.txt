{
  "headerId" : 0,
  "timestamp" : "1970-01-01T00:00:00Z",
  "version" : "version",
  "manufacturer" : "manufacturer",
  "serialNumber" : "serial-number",
  "orderId" : "some-order",
  "orderUpdateId" : 0,
  "lastNodeId" : "some-node",
  "lastNodeSequenceId" : 0,
  "nodeStates" : [ {
    "nodeId" : "node-with-node-position",
    "sequenceId" : 0,
    "released" : true,
    "nodePosition" : {
      "x" : 1.2,
      "y" : 3.4,
      "mapId" : "some-map-id"
    }
  }, {
    "nodeId" : "node-without-node-position",
    "sequenceId" : 0,
    "released" : true
  } ],
  "edgeStates" : [ {
    "edgeId" : "edge-with-trajectory",
    "sequenceId" : 0,
    "released" : true,
    "trajectory" : {
      "degree" : 1.0,
      "knotVector" : [ 0.0 ],
      "controlPoints" : [ {
        "x" : 9.8,
        "y" : 7.6
      } ]
    }
  }, {
    "edgeId" : "edge-without-trajectory",
    "sequenceId" : 0,
    "released" : true
  } ],
  "driving" : true,
  "actionStates" : [ {
    "actionId" : "some-action-id",
    "actionStatus" : "FAILED"
  } ],
  "batteryState" : {
    "batteryCharge" : 50.0,
    "charging" : false
  },
  "operatingMode" : "SEMIAUTOMATIC",
  "errors" : [ {
    "errorType" : "some-error",
    "errorLevel" : "FATAL"
  } ],
  "safetyState" : {
    "eStop" : "AUTOACK",
    "fieldViolation" : false
  },
  "agvPosition" : {
    "x" : 2.1,
    "y" : 5.4,
    "theta" : 0.3,
    "mapId" : "some-map-id",
    "positionInitialized" : true
  },
  "velocity" : { },
  "loads" : [ {
    "boundingBoxReference" : {
      "x" : 0.0,
      "y" : 0.0,
      "z" : 0.0
    },
    "loadDimensions" : {
      "length" : 0.0,
      "width" : 0.0
    }
  }, { } ],
  "information" : [ {
    "infoType" : "some-info",
    "infoLevel" : "DEBUG"
  } ]
}